Yoast/javascript
# Show an orange bullet in the sidebar analysis when a recommended block is removed

## Summary


This PR can be summarized in the following changelog entry:

* [schema-blocks] Adds an orange bullet to the `Analysis` in the side bar when a recommended block is removed.

## Relevant technical choices:

* I added an `enum Blocktype` in `BlockValidationResult.ts`, which holds the different block types (required, recommended, optional, and for the time being, also unknown).
* I added a `blockType` parameter to the `BlockValidationResult` class, with the type being `BlockType`.
* As a result, everywhere where the `BlockVariationResult` class is instantiated, it needs to be passed a `blockType` parameter.
* This is why I also added `unknown` as a value to the enum; I'm passing `unknown` to all instantiations of `BlockValidationResult` which are outside the scope of this PR. We could look into fixing this, but I'm not sure it worth the time investment right now. The thing is that we don't know the block type for many block instances, since the block templates themselves don't pass their own block type (instead, it is listed in `job-posting.block.php` what is the block type for each block).
* The `validateInnerBlocks` function in `innerBlocksValid.ts` now also calls the `findMissingBlocks` function for recommended blocks (it used to call this function only for required blocks).
* I did some renames in the `findMissingBlocks` function to reflect this.
* For each missing/removed block, the `findMissingBlocks` function adds the block type to the `BlockValidationResult`. These results will be used in the sidebar (as the `status` of the `messageData` in `SidebarWarningPresenter.ts`) to show the appropriate warning.

## Test instructions

This PR can be tested by following these steps:

* Add the Yoast Job Posting block.
* See that there are no orange bullets in the `Analysis` in the side bar, because all the recommended blocks are present upon initialization.
* You should see a red bullet saying "Not all required blocks are complete! No job posting schema will be generated for your page." (this is because title, description and location have not yet been filled in).

**Test scenario 7**
* Remove a recommended block, for example the 'Requirements'.
* See an orange bullet and the text: "The 'Requirements' block is recommended but missing.":
    <img width="175" alt="Screenshot 2021-03-23 at 15 21 24" src="https://user-images.githubusercontent.com/20280513/112161436-70bde280-8beb-11eb-99d3-1ebb1f4b9c46.png">

* See that the Requirements block is greyed out and can be added:
    <img width="202" alt="Screenshot 2021-03-23 at 15 21 49" src="https://user-images.githubusercontent.com/20280513/112161492-7fa49500-8beb-11eb-9845-8779a951e405.png">

* Remove another recommended block and see another orange bullet appear (with the right text).
* The red bullet saying "Not all required blocks are complete! No job posting schema will be generated for your page." should still be present.

**Test scenario 8**
* Fill in the title, description and location.
* The red bullet should become green and say: "Good job! All required blocks are complete.".
* You should still see one or more orange bullets corresponding to the recommended blocks you have removed.
* Remove a required block and see that there are now two red bullets: one saying which block is missing, and one saying that not all required blocks are completed.

**Test Schema output**
* Fill in a recommended block and see that it is present in the Schema output.

**Note** that scenarios 7 and 8 both contain a line regarding the color of the emoticon in the pre-publish check; I didn't solve that yet as there is a separate ticket for this task: https://yoast.atlassian.net/browse/P2-789.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  *

## UI changes

## Quality assurance


Fixes https://yoast.atlassian.net/browse/P2-706
